DESIGN-DECISIONS.md

```diff
@@ -0,0 +1,342 @@
+`pgx::aggregate::Aggregate`, described in safe Rust, and pgx handles all the code and SQL generation to expose the 
+aggregate function to the SQL-level.
+
+pgx also provides "unsafe" access to all of Postgres' internals.  Doing so allows pgx, and especially developers, to 
+access all Postgres internals -- especially those not yet wrapped in safe Rust APIs.
+
+The secondary goal, which is not any less important, is an SQL schema generator that programmatically generates 
+extension SQL from the extension Rust sources.  Other than solving the lazy programmer problem, the schema generator 
+ensures certain SQL<-->Postgres<-->Rust invariants are maintained, which includes typing for function arguments and 
+return values.
+
+Finally, pgx wants to improve the day-to-day extension development experience.  `cargo-pgx` is the Cargo plugin that 
+does the heavy lifting of manging Postgres instances for running and testing an extension.
+
+## Non-Goals
+
+pgx does not aim to present safe wrappers around Postgres internals in a one-to-one mapping.  Where it can (and so far 
+has), it endeavors to wrap larger APIs that look and feel like Rust, not Postgres' version of C.  This is an 
+enormous undertaking which will continue for many years.
+
+## Rust Bindings Generation
+
+pgx uses [`bindgen`](https://github.com/rust-lang/rust-bindgen) to generate Rust "bindings" from Postgres' C headers.
+Ultimately, Rust versions of Postgres' internal structs, typedefs, functions definitions, globals, and enums are 
+generated.
+
+The generated bindings are part of the sibling `pgx-pg-sys` crate and are exported through `pgx` as the `::pg_sys` 
+module.  One set of bindings are generated for each supported Postgres version, and the proper one is used based on Rust 
+conditional compilation feature flags.
+
+A bit of post-processing is done to provide some developer conveniences.
+
+First, a graph is built of all Postgres internal structs that appear to be a Postgres `Node`.  The resulting struct 
+definitions are then augmented to `impl Display` (via `pg_sys::nodeToString()`) and to also`impl PgNode`, which is 
+essentially a marker trait.
+
+Additionally, pgx auto-generates a Rust `PgBuiltInOids` enum from the set of `#define xxxOID` defines found in the
+Postgres headers.
+
+pgx reserves the right to generate other convenience types, traits, etc in the future.  Such things may come from user
+requests or from the need to ensure more robust compile-time type checking and safety guarantees.
+
+The bindings are generated via the `build.rs` script in the `pgx-pg-sys` crate.  As some of the Postgres header files 
+are themselves machine-generated, pgx requires its bindings be generated from these headers.  When pgx is managing 
+Postgres instances itself (via `cargo-pgx`), it gets them from there, otherwise the "-server"/"-server-dev" distribution 
+packages must be locally installed.
```

Comment:
Im kinda wondering if that should be re-worded slightly, because some distros use stuff like -server-dev, some use -server-devel, etc. Unless server-dev is the Postgres de-facto way of naming that package